### PR TITLE
Exercise 5 cms

### PR DESCRIPTION
**Pull request by aloparev**
Hello, we have connected CMS as a data source and would like your feedback on that. Also, we didn't get what is the difference between drafted and published data entries in CMS?
Thx in advance!

Edit: Our review can be found [here](https://github.com/Systems-Development-and-Frameworks/bgrakia/pull/12#pullrequestreview-560972686).

### CODE REVIEW

File: backend/src/schema.js [hunk 1]
```diff
@@ -1,39 +1,22 @@
...
-`;
-
-module.exports = typeDefs
\ No newline at end of file
+  let gatewaySchema = stitchSchemas({
+    subschemas: [
+      {schema: graphCmsSchema},
+    ],
+    typeDefs,
+    resolvers,
```
Comment (roschaefer) [inline backend/src/schema.js:18]: You can apply transforms here. I showed how to do that in this commit https://github.com/Systems-Development-and-Frameworks/homework/commit/97b3a0aa4eceb0dfba8cc211aecec776bbaa3753

**Review by roschaefer — CHANGES_REQUESTED** [review]
This is an intermediary feedback, so no grading here. However, your submission looks very, very well. I'm able to decrypt your API keys and start the server. Tests are. See my suggestions for some improvements.

Is it true, that you only seem to use the datasources in your tests? If that is
the case, I would suggest to remove them altogether.
```
$ git grep 'PostsDataSource'
backend/src/DataSources/posts-data-source.js:class PostsDataSource extends DataSource {
backend/src/DataSources/posts-data-source.js:module.exports = {Post, PostsDataSource}
backend/src/Test/context.spec.js:const {PostsDataSource} = require("../DataSources/posts-data-source");
backend/src/Test/context.spec.js:let postsMemory = new PostsDataSource();
backend/src/Test/posts.spec.js:const {Post, PostsDataSource} = require("../DataSources/posts-data-source");
backend/src/Test/posts.spec.js:    postsMemory = new PostsDataSource();
```


Tests are failing for me:
```
yarn run test
yarn run v1.22.10
$ jest
 FAIL  src/Test/context.spec.js
  ● Test suite failed to run

    TypeError: utils.defaultUsers is not iterable

      13 | let usersMemory= new UsersDataSource();
      14 | let postsMemory = new PostsDataSource();
    > 15 | usersMemory.users = [...utils.defaultUsers];
         |                               ^
      16 | postsMemory.posts = [...utils.defaultPosts];
      17 |
      18 |

      at Object.<anonymous> (src/Test/context.spec.js:15:31)

 FAIL  src/Test/posts.spec.js
  ● Test suite failed to run

    TypeError: Server is not a constructor

      24 |
      25 | const testContext = () => decoded
    > 26 | const server = new Server({context: testContext, dataSources: ()=> ({postsDataSrc: postsMemory, usersDataSrc: usersMemory})});
         |                ^
      27 | const {query, mutate } = createTestClient(server);
      28 |
      29 | describe("queries", () => {

      at Object.<anonymous> (src/Test/posts.spec.js:26:16)

 FAIL  src/Test/users.spec.js (7.336 s)
  ● Console

    console.log
      TEST_THECOUNTRYROAD

      at Object.<anonymous> (src/Test/users.spec.js:15:13)

    console.log
      TEST_THECOUNTRYROAD

      at Object.<anonymous> (src/Test/users.spec.js:15:13)

    console.log
      TEST_THECOUNTRYROAD

      at Object.<anonymous> (src/Test/users.spec.js:15:13)

    console.log
      TEST_THECOUNTRYROAD

      at Object.<anonymous> (src/Test/users.spec.js:15:13)

    console.log
      TEST_THECOUNTRYROAD

      at Object.<anonymous> (src/Test/users.spec.js:15:13)

    console.log
      TEST_THECOUNTRYROAD

      at Object.<anonymous> (src/Test/users.spec.js:15:13)

  ● queries › USERS › given users in the database

    expect(received).resolves.toMatchObject(expected)

    - Expected  - 2
    + Received  + 2

    @@ -25,13 +25,13 @@
              "email": "testUser@gmail.com",
              "id": Any<String>,
              "name": "TestUser",
            },
            Object {
    -         "email": "notexistingmail@htw-berlin.de",
    +         "email": "newuser@gmail.com",
              "id": Any<String>,
    -         "name": "Christoph Stach",
    +         "name": "NewUser",
            },
          ],
        },
        "errors": undefined,
      }

      39 |             await expect(query({ query: GET_USERS }))
      40 |             .resolves
    > 41 |             .toMatchObject({
         |              ^
      42 |                 errors: undefined,
      43 |                 data: { people: [
      44 |                     {id:expect.any(String), name:"An", email:"an@gmail.com"},

      at Object.toMatchObject (../node_modules/expect/build/index.js:198:20)
      at Object.<anonymous> (src/Test/users.spec.js:41:14)

  ● queries › USERS › indefinitely nestable query

    expect(received).resolves.toMatchObject(expected)

    - Expected  - 2
    + Received  + 2

    @@ -57,13 +57,13 @@
              "id": Any<String>,
              "name": "TestUser",
              "posts": Array [],
            },
            Object {
    -         "email": "notexistingmail@htw-berlin.de",
    +         "email": "newuser@gmail.com",
              "id": Any<String>,
    -         "name": "Christoph Stach",
    +         "name": "NewUser",
              "posts": Array [],
            },
          ],
        },
        "errors": undefined,

      71 |             await expect(query({query: NESTABLE_QUERY_USERS}))
      72 |            .resolves
    > 73 |            .toMatchObject({
         |             ^
      74 |                errors: undefined,
      75 |                data :{ people:
      76 |                 [

      at Object.toMatchObject (../node_modules/expect/build/index.js:198:20)
      at Object.<anonymous> (src/Test/users.spec.js:73:13)

Test Suites: 3 failed, 3 total
Tests:       2 failed, 4 passed, 6 total
Snapshots:   0 total
Time:        8.772 s, estimated 9 s
Ran all test suites.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

**Comment by roschaefer** [comment]
> Hello, we have connected CMS as a data source and would like your feedback on that. Also, we didn't get what is the difference between drafted and published data entries in CMS?
> Thx in advance!

GraphCMS is a content-management-system (CMS) with a GraphQL API. Having different stages (by default `PUBLISHED` and `DRAFT`) is a typical feature of a CMS. Apparently it's not possible to remove the default stages in GraphCMS or to set it to a single stage like `DEFAULT`. For our use-case we don't use the stages. In general it's an often requested feature to allow editors and administrators to enter data before publishing it. You could e.g. have different URLs for production and staging, staging would receive `PUBLISHED` and `DRAFT` data and production only `PUBLISHED`.

There seems to be a related bug in GraphQL playground regarding unnecessary warnings: https://github.com/graphql/graphql-playground/issues/1299

File: exercises/2/README.md [hunk 2]
```diff
@@ -83,7 +83,7 @@ YourComponent/
 In our lecture of October 21st 2020 we covered online open-source communities and git internals. These topics are not (yet) part of the current homework but will be a task in the upcoming homeworks and/or the final exam. So it's a great opportunity for you to practice the following:
 
 :rocket: Record an online pair-programming.
-- Choose someone who is not a member of your team, preferably some member of an open-source online community but another team member is fine, too.
+- Choose someone who is not a member of your team, preferably some member of an open-source online community but a member of another team is fine, too.
 - Record the pair-programming to [PeerTube](https://joinpeertube.org/en/) or [YouTube](https://www.youtube.com/) and send the link to the video to us
 - Kudos if you publish the video :+1: otherwise mark the video as "unlisted"
```
Comment (MaykAkifovski) [inline exercises/2/README.md:89]: Maybe its better to merge/rebase the changes from the Homework repository in your master/main branch and then to work on the exercise. In this way they wont appear in the PR.

**Review by MaykAkifovski — APPROVED** [review]
lgtm

**Comment by aloparev** [comment]
Final PR is here: https://github.com/Systems-Development-and-Frameworks/CountryRoads/pull/32#issue-552277014

**Pull request closed without merging**